test/common/upstream/eds_speed_test.cc

```diff
@@ -39,16 +40,24 @@ namespace Upstream {
 
 class EdsSpeedTest {
 public:
-  EdsSpeedTest(State& state)
-      : state_(state),
+  EdsSpeedTest(State& state, bool use_unified_mux)
```

Comment:
Nice!
I'm assuming there won't be much performance difference here between unified/non-unified. Can you post the results in a comment, so we'll have them documented as part of this PR?

Reply:
The full test output:
```
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
priorityAndLocalityWeighted/0/1/0            40.5 ms         40.3 ms            1
priorityAndLocalityWeighted/1/1/0            27.6 ms         27.5 ms            1
priorityAndLocalityWeighted/0/8/0            27.7 ms         27.6 ms            1
priorityAndLocalityWeighted/1/8/0            27.7 ms         27.5 ms            1
priorityAndLocalityWeighted/0/64/0           27.2 ms         27.1 ms            1
priorityAndLocalityWeighted/1/64/0           28.3 ms         28.2 ms            1
priorityAndLocalityWeighted/0/512/0          27.6 ms         27.5 ms            1
priorityAndLocalityWeighted/1/512/0          27.8 ms         27.7 ms            1
priorityAndLocalityWeighted/0/4096/0         27.7 ms         27.6 ms            1
priorityAndLocalityWeighted/1/4096/0         27.7 ms         27.6 ms            1
priorityAndLocalityWeighted/0/32768/0        28.0 ms         27.9 ms            1
priorityAndLocalityWeighted/1/32768/0        27.5 ms         27.4 ms            1
priorityAndLocalityWeighted/0/100000/0       29.0 ms         28.9 ms            1
priorityAndLocalityWeighted/1/100000/0       27.7 ms         27.5 ms            1
priorityAndLocalityWeighted/0/1/1            28.2 ms         28.1 ms            1
priorityAndLocalityWeighted/1/1/1            28.0 ms         27.9 ms            1
priorityAndLocalityWeighted/0/8/1            27.6 ms         27.5 ms            1
priorityAndLocalityWeighted/1/8/1            28.3 ms         28.2 ms            1
priorityAndLocalityWeighted/0/64/1           28.7 ms         28.6 ms            1
priorityAndLocalityWeighted/1/64/1           27.7 ms         27.6 ms            1
priorityAndLocalityWeighted/0/512/1          28.0 ms         27.9 ms            1
priorityAndLocalityWeighted/1/512/1          30.0 ms         29.8 ms            1
priorityAndLocalityWeighted/0/4096/1         28.6 ms         28.4 ms            1
priorityAndLocalityWeighted/1/4096/1         27.1 ms         27.0 ms            1
priorityAndLocalityWeighted/0/32768/1        27.6 ms         27.5 ms            1
priorityAndLocalityWeighted/1/32768/1        27.5 ms         27.4 ms            1
priorityAndLocalityWeighted/0/100000/1       26.9 ms         26.8 ms            1
priorityAndLocalityWeighted/1/100000/1       26.9 ms         26.8 ms            1
duplicateUpdate/1/0                          27.8 ms         27.7 ms            1
duplicateUpdate/8/0                          32.4 ms         32.2 ms            1
duplicateUpdate/64/0                         26.9 ms         26.8 ms            1
duplicateUpdate/512/0                        27.6 ms         27.5 ms            1
duplicateUpdate/4096/0                       29.3 ms         29.2 ms            1
duplicateUpdate/32768/0                      27.1 ms         27.0 ms            1
duplicateUpdate/100000/0                     26.9 ms         26.8 ms            1
duplicateUpdate/1/1                          25.7 ms         25.6 ms            1
duplicateUpdate/8/1                          27.4 ms         27.3 ms            1
duplicateUpdate/64/1                         27.8 ms         27.7 ms            1
duplicateUpdate/512/1                        29.9 ms         29.7 ms            1
duplicateUpdate/4096/1                       28.5 ms         28.4 ms            1
duplicateUpdate/32768/1                      28.4 ms         28.3 ms            1
duplicateUpdate/100000/1                     27.9 ms         27.8 ms            1
healthOnlyUpdate/1/0                         26.8 ms         26.7 ms            1
healthOnlyUpdate/8/0                         26.7 ms         26.6 ms            1
healthOnlyUpdate/64/0                        27.0 ms         26.9 ms            1
healthOnlyUpdate/512/0                       25.6 ms         25.5 ms            1
healthOnlyUpdate/4096/0                      27.8 ms         27.7 ms            1
healthOnlyUpdate/32768/0                     26.9 ms         26.8 ms            1
healthOnlyUpdate/100000/0                    29.2 ms         29.1 ms            1
healthOnlyUpdate/1/1                         30.1 ms         30.0 ms            1
healthOnlyUpdate/8/1                         30.4 ms         30.3 ms            1
healthOnlyUpdate/64/1                        30.4 ms         30.3 ms            1
healthOnlyUpdate/512/1                       29.8 ms         29.7 ms            1
healthOnlyUpdate/4096/1                      29.7 ms         29.5 ms            1
healthOnlyUpdate/32768/1                     29.6 ms         29.5 ms            1
healthOnlyUpdate/100000/1                    29.4 ms         29.3 ms            1
```

Test results with names ending with a `/1` are  for unified mux.

Reply:
So, I would expect almost no difference given things are functionally the same. But these results indicate something more interesting. How come there is only a single iteration?

Reply:
Ran without a wrapper and `benchmark_min_time=2`:
```
Run on (12 X 4400 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 256 KiB (x6)
  L3 Unified 12288 KiB (x1)
Load Average: 0.85, 1.70, 4.68
---------------------------------------------------------------------------------
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
priorityAndLocalityWeighted/0/1/0            2.17 ms         2.16 ms         1243
priorityAndLocalityWeighted/1/1/0            1.99 ms         1.99 ms         1231
priorityAndLocalityWeighted/0/8/0            2.06 ms         2.05 ms         1326
priorityAndLocalityWeighted/1/8/0            2.01 ms         2.00 ms         1324
priorityAndLocalityWeighted/0/64/0           2.11 ms         2.10 ms         1288
priorityAndLocalityWeighted/1/64/0           2.06 ms         2.05 ms         1317
priorityAndLocalityWeighted/0/512/0          3.24 ms         3.23 ms          856
priorityAndLocalityWeighted/1/512/0          2.81 ms         2.80 ms          969
priorityAndLocalityWeighted/0/4096/0         12.5 ms         12.5 ms          225
priorityAndLocalityWeighted/1/4096/0         9.34 ms         9.32 ms          301
priorityAndLocalityWeighted/0/32768/0         100 ms         99.8 ms           28
priorityAndLocalityWeighted/1/32768/0        75.1 ms         74.9 ms           37
priorityAndLocalityWeighted/0/100000/0        322 ms          321 ms            8
priorityAndLocalityWeighted/1/100000/0        243 ms          242 ms           11
priorityAndLocalityWeighted/0/1/1            1.96 ms         1.95 ms         1384
priorityAndLocalityWeighted/1/1/1            2.01 ms         2.00 ms         1387
priorityAndLocalityWeighted/0/8/1            1.98 ms         1.98 ms         1393
priorityAndLocalityWeighted/1/8/1            1.97 ms         1.96 ms         1374
priorityAndLocalityWeighted/0/64/1           2.19 ms         2.18 ms         1229
priorityAndLocalityWeighted/1/64/1           2.07 ms         2.06 ms         1306
priorityAndLocalityWeighted/0/512/1          3.87 ms         3.86 ms          721
priorityAndLocalityWeighted/1/512/1          3.00 ms         2.99 ms          917
priorityAndLocalityWeighted/0/4096/1         17.0 ms         17.0 ms          163
priorityAndLocalityWeighted/1/4096/1         10.9 ms         10.9 ms          255
priorityAndLocalityWeighted/0/32768/1         137 ms          137 ms           20
priorityAndLocalityWeighted/1/32768/1        87.8 ms         87.6 ms           32
priorityAndLocalityWeighted/0/100000/1        435 ms          434 ms            6
priorityAndLocalityWeighted/1/100000/1        284 ms          283 ms           10
duplicateUpdate/1/0                          1.96 ms         1.96 ms         1387
duplicateUpdate/8/0                          1.98 ms         1.98 ms         1343
duplicateUpdate/64/0                         2.16 ms         2.15 ms         1270
duplicateUpdate/512/0                        3.57 ms         3.56 ms          789
duplicateUpdate/4096/0                       16.1 ms         16.0 ms          174
duplicateUpdate/32768/0                       137 ms          136 ms           21
duplicateUpdate/100000/0                      455 ms          454 ms            6
duplicateUpdate/1/1                          1.99 ms         1.98 ms         1383
duplicateUpdate/8/1                          2.00 ms         2.00 ms         1350
duplicateUpdate/64/1                         2.21 ms         2.20 ms         1185
duplicateUpdate/512/1                        3.91 ms         3.90 ms          718
duplicateUpdate/4096/1                       19.1 ms         19.0 ms          138
duplicateUpdate/32768/1                       162 ms          162 ms           17
duplicateUpdate/100000/1                      533 ms          532 ms            5
healthOnlyUpdate/1/0                         1.96 ms         1.96 ms         1384
healthOnlyUpdate/8/0                         1.98 ms         1.98 ms         1374
healthOnlyUpdate/64/0                        2.16 ms         2.15 ms         1260
healthOnlyUpdate/512/0                       3.62 ms         3.61 ms          764
healthOnlyUpdate/4096/0                      16.6 ms         16.6 ms          168
healthOnlyUpdate/32768/0                      145 ms          144 ms           19
healthOnlyUpdate/100000/0                     491 ms          489 ms            5
healthOnlyUpdate/1/1                         1.99 ms         1.98 ms         1372
healthOnlyUpdate/8/1                         2.01 ms         2.01 ms         1347
healthOnlyUpdate/64/1                        2.23 ms         2.22 ms         1228
healthOnlyUpdate/512/1                       3.99 ms         3.98 ms          700
healthOnlyUpdate/4096/1                      19.7 ms         19.6 ms          125
healthOnlyUpdate/32768/1                      169 ms          169 ms           17
healthOnlyUpdate/100000/1                     566 ms          565 ms            5
```

Reply:
I haven't actually tried profiling this, but I suspect the performance hit is here: https://github.com/envoyproxy/envoy/blob/main/source/common/config/xds_mux/sotw_subscription_state.cc#L62. We effectively are decoding a resource twice: once to update ttl, and then again in `WatchMap::onConfigUpdate`, before calling subscriptions's onConfigUpdate. I'll see if I can skip the second update...

Reply:
Yeah, that was it. I'll clean it up and check if delta needs a similar fix (I think it does). Here's the speed test results with the issue mentioned above fixed:

```
---------------------------------------------------------------------------------
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
priorityAndLocalityWeighted/0/1/0            2.10 ms         2.10 ms         1139
priorityAndLocalityWeighted/1/1/0            2.00 ms         2.00 ms         1228
priorityAndLocalityWeighted/0/8/0            2.11 ms         2.10 ms         1350
priorityAndLocalityWeighted/1/8/0            2.07 ms         2.06 ms         1339
priorityAndLocalityWeighted/0/64/0           2.11 ms         2.11 ms         1284
priorityAndLocalityWeighted/1/64/0           2.07 ms         2.06 ms         1302
priorityAndLocalityWeighted/0/512/0          3.29 ms         3.28 ms          836
priorityAndLocalityWeighted/1/512/0          2.81 ms         2.80 ms          972
priorityAndLocalityWeighted/0/4096/0         12.5 ms         12.5 ms          223
priorityAndLocalityWeighted/1/4096/0         9.34 ms         9.31 ms          302
priorityAndLocalityWeighted/0/32768/0         101 ms          101 ms           27
priorityAndLocalityWeighted/1/32768/0        76.1 ms         75.9 ms           37
priorityAndLocalityWeighted/0/100000/0        327 ms          326 ms            8
priorityAndLocalityWeighted/1/100000/0        246 ms          245 ms           11
priorityAndLocalityWeighted/0/1/1            1.96 ms         1.95 ms         1380
priorityAndLocalityWeighted/1/1/1            1.95 ms         1.94 ms         1366
priorityAndLocalityWeighted/0/8/1            1.98 ms         1.97 ms         1362
priorityAndLocalityWeighted/1/8/1            1.97 ms         1.96 ms         1362
priorityAndLocalityWeighted/0/64/1           2.12 ms         2.11 ms         1286
priorityAndLocalityWeighted/1/64/1           2.06 ms         2.05 ms         1311
priorityAndLocalityWeighted/0/512/1          3.27 ms         3.26 ms          816
priorityAndLocalityWeighted/1/512/1          2.86 ms         2.85 ms          952
priorityAndLocalityWeighted/0/4096/1         12.7 ms         12.7 ms          221
priorityAndLocalityWeighted/1/4096/1         9.42 ms         9.39 ms          297
priorityAndLocalityWeighted/0/32768/1         101 ms          101 ms           28
priorityAndLocalityWeighted/1/32768/1        77.0 ms         76.8 ms           37
priorityAndLocalityWeighted/0/100000/1        323 ms          322 ms            8
priorityAndLocalityWeighted/1/100000/1        246 ms          245 ms           11
duplicateUpdate/1/0                          1.97 ms         1.96 ms         1390
duplicateUpdate/8/0                          1.99 ms         1.98 ms         1373
duplicateUpdate/64/0                         2.16 ms         2.15 ms         1226
duplicateUpdate/512/0                        3.57 ms         3.56 ms          769
duplicateUpdate/4096/0                       16.7 ms         16.6 ms          175
duplicateUpdate/32768/0                       139 ms          139 ms           20
duplicateUpdate/100000/0                      463 ms          462 ms            6
duplicateUpdate/1/1                          2.00 ms         1.99 ms         1364
duplicateUpdate/8/1                          2.06 ms         2.05 ms         1339
duplicateUpdate/64/1                         2.20 ms         2.19 ms         1241
duplicateUpdate/512/1                        3.65 ms         3.64 ms          776
duplicateUpdate/4096/1                       16.5 ms         16.5 ms          171
duplicateUpdate/32768/1                       139 ms          139 ms           20
duplicateUpdate/100000/1                      459 ms          458 ms            6
healthOnlyUpdate/1/0                         1.99 ms         1.98 ms         1367
healthOnlyUpdate/8/0                         2.00 ms         1.99 ms         1342
healthOnlyUpdate/64/0                        2.17 ms         2.16 ms         1252
healthOnlyUpdate/512/0                       3.64 ms         3.63 ms          762
healthOnlyUpdate/4096/0                      17.2 ms         17.1 ms          167
healthOnlyUpdate/32768/0                      148 ms          148 ms           19
healthOnlyUpdate/100000/0                     491 ms          490 ms            5
healthOnlyUpdate/1/1                         2.00 ms         1.99 ms         1353
healthOnlyUpdate/8/1                         2.03 ms         2.02 ms         1271
healthOnlyUpdate/64/1                        2.26 ms         2.25 ms         1234
healthOnlyUpdate/512/1                       3.73 ms         3.72 ms          739
healthOnlyUpdate/4096/1                      17.5 ms         17.5 ms          163
healthOnlyUpdate/32768/1                      148 ms          148 ms           19
healthOnlyUpdate/100000/1                     497 ms          495 ms            5
```

Reply:
eds speed test results with the latest changes:

```
priorityAndLocalityWeighted/0/1/0            2.10 ms         2.09 ms         1278
priorityAndLocalityWeighted/1/1/0            2.07 ms         2.07 ms         1320
priorityAndLocalityWeighted/0/8/0            2.22 ms         2.21 ms         1314
priorityAndLocalityWeighted/1/8/0            2.17 ms         2.16 ms         1253
priorityAndLocalityWeighted/0/64/0           2.20 ms         2.19 ms         1225
priorityAndLocalityWeighted/1/64/0           2.12 ms         2.12 ms         1286
priorityAndLocalityWeighted/0/512/0          3.36 ms         3.35 ms          829
priorityAndLocalityWeighted/1/512/0          3.09 ms         3.07 ms          927
priorityAndLocalityWeighted/0/4096/0         13.1 ms         13.1 ms          212
priorityAndLocalityWeighted/1/4096/0         9.86 ms         9.83 ms          284
priorityAndLocalityWeighted/0/32768/0         103 ms          103 ms           27
priorityAndLocalityWeighted/1/32768/0        78.2 ms         78.0 ms           36
priorityAndLocalityWeighted/0/100000/0        329 ms          328 ms            8
priorityAndLocalityWeighted/1/100000/0        244 ms          244 ms           11
priorityAndLocalityWeighted/0/1/1            1.90 ms         1.89 ms         1474
priorityAndLocalityWeighted/1/1/1            1.89 ms         1.88 ms         1481
priorityAndLocalityWeighted/0/8/1            1.98 ms         1.97 ms         1458
priorityAndLocalityWeighted/1/8/1            2.03 ms         2.02 ms         1429
priorityAndLocalityWeighted/0/64/1           2.36 ms         2.32 ms         1092
priorityAndLocalityWeighted/1/64/1           2.34 ms         2.33 ms         1294
priorityAndLocalityWeighted/0/512/1          3.22 ms         3.21 ms          866
priorityAndLocalityWeighted/1/512/1          2.80 ms         2.79 ms          971
priorityAndLocalityWeighted/0/4096/1         12.8 ms         12.7 ms          217
priorityAndLocalityWeighted/1/4096/1         9.66 ms         9.63 ms          292
priorityAndLocalityWeighted/0/32768/1         101 ms          100 ms           28
priorityAndLocalityWeighted/1/32768/1        75.9 ms         75.7 ms           36
priorityAndLocalityWeighted/0/100000/1        316 ms          316 ms            8
priorityAndLocalityWeighted/1/100000/1        247 ms          246 ms           11
duplicateUpdate/1/0                          1.93 ms         1.92 ms         1453
duplicateUpdate/8/0                          1.96 ms         1.95 ms         1382
duplicateUpdate/64/0                         2.13 ms         2.12 ms         1273
duplicateUpdate/512/0                        3.51 ms         3.50 ms          790
duplicateUpdate/4096/0                       16.4 ms         16.4 ms          169
duplicateUpdate/32768/0                       139 ms          139 ms           20
duplicateUpdate/100000/0                      463 ms          462 ms            6
duplicateUpdate/1/1                          1.93 ms         1.93 ms         1376
duplicateUpdate/8/1                          1.96 ms         1.96 ms         1396
duplicateUpdate/64/1                         2.12 ms         2.12 ms         1284
duplicateUpdate/512/1                        3.49 ms         3.48 ms          776
duplicateUpdate/4096/1                       16.3 ms         16.2 ms          170
duplicateUpdate/32768/1                       140 ms          140 ms           20
duplicateUpdate/100000/1                      459 ms          458 ms            6
healthOnlyUpdate/1/0                         1.94 ms         1.93 ms         1457
healthOnlyUpdate/8/0                         1.96 ms         1.96 ms         1377
healthOnlyUpdate/64/0                        2.13 ms         2.13 ms         1187
healthOnlyUpdate/512/0                       3.60 ms         3.59 ms          781
healthOnlyUpdate/4096/0                      17.0 ms         16.9 ms          165
healthOnlyUpdate/32768/0                      148 ms          148 ms           19
healthOnlyUpdate/100000/0                     489 ms          488 ms            6
healthOnlyUpdate/1/1                         1.97 ms         1.96 ms         1443
healthOnlyUpdate/8/1                         1.94 ms         1.93 ms         1382
healthOnlyUpdate/64/1                        2.11 ms         2.10 ms         1288
healthOnlyUpdate/512/1                       3.53 ms         3.52 ms          786
healthOnlyUpdate/4096/1                      16.8 ms         16.8 ms          165
healthOnlyUpdate/32768/1                      149 ms          149 ms           19
healthOnlyUpdate/100000/1                     489 ms          488 ms            6
```